src/OpenFeature/HookData.cs

```diff
@@ -0,0 +1,102 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using OpenFeature.Model;
+
+namespace OpenFeature
+{
+    /// <summary>
+    /// A key-value collection of strings to objects used for passing data between hook stages.
+    /// <para>
+    /// This collection is scoped to a single evaluation for a single hook. Each hook stage for the evaluation
+    /// will share the same <see cref="HookData"/>.
+    /// </para>
+    /// <para>
+    /// This collection is intended for use only during the execution of individual hook stages, a reference
+    /// to the collection should not be retained.
+    /// </para>
+    /// <para>
+    /// This collection is not thread-safe.
+    /// </para>
+    /// </summary>
+    /// <seealso href="https://github.com/open-feature/spec/blob/main/specification/sections/04-hooks.md#46-hook-data"/>
+    public sealed class HookData
+    {
+        private readonly Dictionary<string, object> _data = [];
+
+        /// <summary>
+        /// Set the key to the given value.
+        /// </summary>
+        /// <param name="key">The key for the value</param>
+        /// <param name="value">The value to set</param>
+        public void Set(string key, object value)
+        {
+            this._data[key] = value;
+        }
```

Comment:
Maybe we could return `HookData` to allow `hookData.Set("a", "aaa").Set("b", "bbb")` fluent set expression

```suggestion
        public HookData Set(string key, object value)
        {
            this._data[key] = value;
            return this;
        }
```

Reply:
It is a great suggestion. Thanks for raising it 👍

Reply:
Added.